### PR TITLE
Implement MatrixFreeTools::compute_matrix()

### PR DESCRIPTION
**Pull request by peterrum**
Needed for https://github.com/dealii/dealii/pull/11254 and something I wanted to add for a very long time.

### CODE REVIEW

File: include/deal.II/matrix_free/tools.h [hunk 1]
```diff
@@ -509,6 +567,153 @@ namespace MatrixFreeTools
...
+
+        unsigned int const dofs_per_cell = integrator.dofs_per_cell;
+
+        for (auto cell = range.first; cell < range.second; ++cell)
+          {
+            unsigned int const n_filled_lanes =
+              matrix_free.n_active_entries_per_cell_batch(cell);
+
+            FullMatrix<TrilinosScalar>
+              matrices[VectorizedArray<double>::size()];
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:611]: Should this be
```suggestion
              matrices[VectorizedArrayType::size()];
```

File: include/deal.II/matrix_free/tools.h [hunk 2]
```diff
@@ -509,6 +567,153 @@ namespace MatrixFreeTools
...
+
+            std::fill_n(matrices,
+                        VectorizedArray<double>::size(),
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:614]:
```suggestion
                        VectorizedArrayType::size(),
```

File: include/deal.II/matrix_free/tools.h [hunk 3]
```diff
@@ -509,6 +567,153 @@ namespace MatrixFreeTools
...
+                        FullMatrix<TrilinosScalar>(dofs_per_cell,
+                                                   dofs_per_cell));
+
+            integrator.reinit(cell);
+
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+              {
+                for (unsigned int i = 0; i < integrator.dofs_per_cell; ++i)
+                  integrator.begin_dof_values()[i] =
+                    static_cast<double>(i == j);
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:624]:
```suggestion
                    static_cast<Number>(i == j);
```

File: include/deal.II/matrix_free/tools.h [hunk 4]
```diff
@@ -509,6 +567,153 @@ namespace MatrixFreeTools
...
+                    matrices[v](i, j) = integrator.begin_dof_values()[i][v];
+              }
+
+
+            // finally assemble local matrices into global matrix
+            for (unsigned int v = 0; v < n_filled_lanes; v++)
+              {
+                auto cell_v = matrix_free.get_cell_iterator(cell, v);
+
+                std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:639]: I am not sure whether it matters at this level, but I tend to prefer moving the allocation of `dof_indices` (and `temp` below) outside of this loop (i.e., line 604) so the memory allocation gets amortized a bit. (Now, to be honest, at least for the case of complicated constraints there is a memory allocation happening inside `constraints.distribute_local_to_global`, so it won't matter in the grand scheme.)

**Review by kronbichler — COMMENTED** [review]
Looks good, except for a few comments on non-`double` data types.

**Comment by peterrum** [comment]
@kronbichler Thanks for the review. I have made the changes!

**Review by kronbichler — APPROVED** [review]

**Comment by peterrum** [comment]
@kronbichler I have made some additional stylistic changes. Now, this PR is ready from my side. Once this PR, #11307 and #11308 are merged, I can rebase #11254 (step-75).

**Review by kronbichler — APPROVED** [review]

**Pull request merged**